server/src/tracing.ts

```diff
@@ -0,0 +1,187 @@
+/**
+ * Tools for instrumenting code with traces. This is mainly designed to support
+ * Sentry's tracing features, but could be expanded support other approaches,
+ * like OpenTelemetry.
+ */
+
+import * as Sentry from "@sentry/node";
+import type { Span, SpanStatusType } from "@sentry/tracing";
+import type { SpanContext } from "@sentry/types";
+export type { Span } from "@sentry/tracing";
+
+interface SpanOptions extends SpanContext {
+  parentSpan?: Span;
+  timeout?: number;
+}
+
+/**
+ * Start a tracing span. The returned span should be explicitly ended with
+ * `finishSpan`. The created span will be a child of whatever span is currently
+ * active (and then become the current span itself), or if there is no current
+ * span or transaction, this will start one for you.
+ *
+ * Alternatively, can explicitly pass an actual span object to be the parent:
+ * `startSpan({ parentSpan: yourSpan })`. In this case, the new span won't
+ * automatically become the global "current" span.
+ *
+ * Spans created this way will be automatically canceled when their transaction
+ * finishes (Sentry will drop any unfinished spans). Alternatively, you can
+ * set `timeout` to a number of milliseconds, and the span will be canceled
+ * after that time (this should prevent you from accidentally leaving a span
+ * open forever). Canceled spans have a non-ok status set and a `cancel` tag
+ * with a reason.
+ *
+ * More on why spans need canceling:
+ * https://github.com/getsentry/sentry-javascript/issues/4165#issuecomment-971424754
+ */
+export function startSpan(options: SpanOptions): Span {
+  let { parentSpan, timeout, ...spanOptions } = options;
+
+  let scope;
+  if (!parentSpan) {
+    scope = Sentry.getCurrentHub().getScope();
+    parentSpan = scope.getSpan();
+  }
+
+  let newSpan: Span;
+  if (parentSpan) {
+    newSpan = parentSpan.startChild(spanOptions);
+  } else {
+    newSpan = Sentry.startTransaction(spanOptions as any);
+  }
+
+  // If we retrieved the span from the scope, update the scope.
+  if (scope) {
+    scope.setSpan(newSpan);
+  }
+
+  if (timeout) {
+    setTimeout(() => {
+      cancelSpan(newSpan, "deadline_exceeded");
+    }, timeout).unref();
+  } else if (newSpan.transaction && newSpan.transaction !== newSpan) {
+    // FIXME: newer Sentry has an event for this: "finishTransaction" emitted
+    // on the hub's client:
+    // - Code: https://github.com/getsentry/sentry-javascript/blob/ba99e7cdf725725e5a1b99e9d814353dbb3ae2b6/packages/core/src/tracing/transaction.ts#L144-L147
+    // - Feature: https://github.com/getsentry/sentry-javascript/issues/7262
+    const _finishTransaction = newSpan.transaction.finish;
+    newSpan.transaction.finish = function finish(...args) {
+      cancelSpan(newSpan, "cancelled", "did_not_finish");
+      return _finishTransaction.call(this, ...args);
+    };
```

Comment:
More dumb lessons: the transaction event is on the client, which belongs to the hub. But the hub is stored on the transaction as a `_hub` property that you cannot get publicly!

So we can use the event if we grab a private property that might change, or we can keep this monkey-patching mess, which in practical terms is actually more stable. Sigh.